### ISSUE TITLE
[SDK 45.0.6 RN 0.69.3] - [Android] - package expo.modules does not exist in MainApplication.java

### ISSUE THREAD

**Issue by 19424056** [expo/expo#18433]
### Summary

I can't compile android after upgrade to expo sdk 45
I have tested for ios, works well

Existing Apps
Setting up the expo package
https://github.com/expo/fyi/blob/main/expo-modules-migration.md

error appears in log :
> Task :app:compressReleaseAssets
Execution optimizations have been disabled for task ':app:compressReleaseAssets' to ensure correctness due to the following reasons:
- Gradle detected a problem with the following location: '/Users/momo/Documents/test/test_expo/android/app/build/intermediates/assets/release/mergeReleaseAssets'. Reason: Task ':app:compressReleaseAssets' uses this output of task ':app:copyReleaseBundledJs' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.

> Task :app:compileReleaseJavaWithJavac FAILED
/Users/momo/Documents/test/test_expo/android/app/src/main/java/com/test_expo/MainApplication.java:27: error: package expo.modules does not exist
import expo.modules.ApplicationLifecycleDispatcher;
^
/Users/momo/Documents/test/test_expo/android/app/src/main/java/com/test_expo/MainApplication.java:28: error: package expo.modules does not exist
import expo.modules.ReactNativeHostWrapper;
^
/Users/momo/Documents/test/test_expo/android/app/src/main/java/com/test_expo/MainActivity.java:8: error: package expo.modules does not exist
import expo.modules.ReactActivityDelegateWrapper;
^
/Users/momo/Documents/test/test_expo/android/app/src/main/java/com/test_expo/MainApplication.java:31: error: cannot find symbol
private final ReactNativeHost mReactNativeHost = new ReactNativeHostWrapper(
^
symbol: class ReactNativeHostWrapper
location: class MainApplication
/Users/momo/Documents/test/test_expo/android/app/src/main/java/com/test_expo/MainApplication.java:85: error: cannot find symbol
ApplicationLifecycleDispatcher.onApplicationCreate(this);
^
symbol: variable ApplicationLifecycleDispatcher
location: class MainApplication
/Users/momo/Documents/test/test_expo/android/app/src/main/java/com/test_expo/MainApplication.java:91: error: cannot find symbol
ApplicationLifecycleDispatcher.onConfigurationChanged(this, newConfig);
^
symbol: variable ApplicationLifecycleDispatcher
location: class MainApplication
/Users/momo/Documents/test/test_expo/android/app/src/main/java/com/test_expo/MainActivity.java:38: error: cannot find symbol
return new ReactActivityDelegateWrapper(
^
symbol: class ReactActivityDelegateWrapper
location: class MainActivity
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
7 errors
w: Detected multiple Kotlin daemon sessions at build/kotlin/sessions

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileReleaseJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org/

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.4.2/userguide/command_line_interface.html#sec:command_line_warnings

Execution optimizations have been disabled for 2 invalid unit(s) of work during this build to ensure correctness.
Please consult deprecation warnings for more details.

BUILD FAILED in 4m 15s
773 actionable tasks: 727 executed, 46 up-to-date
Dealing with validation problems

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

0.45.6

### Environment

System:
    OS: macOS 12.4
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
    Memory: 26.73 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.15.1 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 8.11.0 - /usr/local/bin/npm
    Watchman: 2022.07.04.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    Android SDK: Not Found
  IDEs:
    Android Studio: 2021.2 AI-212.5712.43.2112.8609683
    Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.15 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.0.0 => 18.0.0
    react-native: 0.69.3 => 0.69.3
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
    
    
    npmPackages:
      react (18.0.0)
      react-native (0.69.3)
      expo 45.0.6
      expo-application (4.1.0)
      expo-constants (13.1.1)
      expo-file-system (14.0.0)
      expo-font (10.1.0)
      expo-image-loader (3.2.0)
      expo-image-manipulator (10.3.1)
      expo-in-app-purchases (13.0.0)
      expo-location (14.2.2)
      expo-modules-core (0.9.2)
      
      AGP version is 7.4.2-all
        classpath("com.android.tools.build:gradle:7.1.1")
        classpath("com.facebook.react:react-native-gradle-plugin")
        classpath("de.undercouch:gradle-download-task:5.0.1")

### Reproducible demo

Just upgrade to expo 45.0.6

https://github.com/expo/fyi/blob/main/expo-modules-migration.md
![image](https://user-images.githubusercontent.com/60703498/182099950-164e6286-c13a-490e-9f41-c42ffda4e547.png)

**Comment by buidinhmanh14799** [comment]
same issue . please help me @Kudo

**Comment by Kudo** [comment]
hi there! it's hard for us to investigate the issue without minimal reproducible example repo. please have a minimal example repo and i'm happy to investigate and reopen the issue.

**Comment by 2MO5** [comment]
Hello there!

Did you find a solution or a way around this mess?

**Comment by ardyfeb** [comment]
Getting this issue on `expo: ^49.0.2` too

**Comment by saif-o99** [comment]
Got the same with version `49.0.2` and RN `0.72.1` @Kudo

**Comment by qusaieilouti99** [comment]
I got the same issue here on 49.0.2 while updating to latest react native version

**Comment by saif-o99** [comment]
@ardyfeb did you manage to fix it?

**Comment by ardyfeb** [comment]
@saif-o99 No, that error still blocks me

**Comment by saif-o99** [comment]
Now its working, i created new react native project with expo and started to compare android folder files with mine. i dont know what exactly fixed the issue but what think it fixed the issue is inside settings.gradle

before:
````
apply from: new File(["node", "--print", "require.resolve('expo/package.json')"].execute().text.trim(), "../scripts/autolinking.gradle");
useExpoModules()

include ':app'
includeBuild(new File(["node", "--print", "require.resolve('react-native-gradle-plugin/package.json')"].execute(null, rootDir).text.trim()).getParentFile())

````

after:
````
apply from: new File(["node", "--print", "require.resolve('expo/package.json')"].execute(null, rootDir).text.trim(), "../scripts/autolinking.gradle")
useExpoModules()

include ':app'
includeBuild('../node_modules/@react-native/gradle-plugin')

````

**Comment by Bardiamist** [comment]
Other things in [configuration](https://docs.expo.dev/bare/installing-expo-modules/#configuration-for-android) also changed.

@brentvatne I expected if something changed then it should be mentioned in https://blog.expo.dev/expo-sdk-49-c6d398cdf740
Ideally we need migration guide:)

**Comment by alerighi** [comment]
I've the same issue after upgrading to SDK 49. I cannot manage to make the build on Android work. I copied the files from the template-bare-minimum project and they seem identical to me. 

Why upgrading to a new version is always a pain? Everything seems to unstable and breaks for unexpected things for no apparent reason...

**Comment by phil-hudson** [comment]
hey - did anyone manage to fix this?

**Comment by bgdawie** [comment]
@phil-hudson 

Yes, I managed to resolve the expo.modules not found issue in MainApplication.java. Here's what worked for me:

Switch from npm to Yarn: 
I initially faced auto-linking issues with npm. After removing package-lock.json and switching to Yarn for dependency management, the auto-linking problem was resolved.

Expo Version: 
Ensure that you're using the correct version of Expo. In my case, I manually downgraded to Expo 49.0.16.

Reinstallation of Node Modules: 
After switching to Yarn, I did a clean installation of the node modules.

Checking Configuration: Double-check your MainApplication.java and other configuration files to ensure that there are no path or naming issues with the expo.modules.

Clean and Rebuild: Finally, clean your project and rebuild it. This can help in resolving any lingering issues from the previous builds.

In anycase this is more of a workaround, im not sure why the autolinking with npm isnt working,

my npm version is 8.19.4
my node version is 16.20.2

hope this helps

**Comment by DawiePritchard** [comment]
@phil-hudson

Yes, I managed to resolve the expo.modules not found issue in MainApplication.java. Here's what worked for me:

Switch from npm to Yarn:
I initially faced auto-linking issues with npm. After removing package-lock.json and switching to Yarn for dependency management, the auto-linking problem was resolved.

Expo Version:
Ensure that you're using the correct version of Expo. In my case, I manually downgraded to Expo 49.0.16.

Reinstallation of Node Modules:
After switching to Yarn, I did a clean installation of the node modules.

Checking Configuration: Double-check your MainApplication.java and other configuration files to ensure that there are no path or naming issues with the expo.modules.

Clean and Rebuild: Finally, clean your project and rebuild it. This can help in resolving any lingering issues from the previous builds.

In anycase this is more of a workaround, im not sure why the autolinking with npm isnt working,

my npm version is 8.19.4
my node version is 16.20.2

hope this helps

**Comment by liamjones** [comment]
I've just hit this upgrading from SDK 48 to 49 in a bare RN project too.

48's auto-linking was working fine, 49 breaks and doesn't appear to add a dynamic `implementation project(':expo')` to the app dependencies. I've just done it manually for now.

**Comment by liamjones** [comment]
Okay, I've discovered why the upgrade from Expo SDK 48 to 49 broke auto-linking for us. I'm sharing the explanation/solution in case anyone else runs into it:

SDK 49 is trying to be smart about whether it should auto-link or not and makes some, in my opinion, naive assumptions about the setup of your `android/settings.gradle` and `ios/Podfile` files.

The `expo` package's `react-native.config.js` decides whether to auto-link or not based on a regex run against the files mentioned above: https://github.com/expo/expo/blob/0dbdb7248975901d46b943623b2891f592766735/packages/expo/react-native.config.js#L44-L54

SDK 48 didn't do this, it always auto-linked: https://github.com/expo/expo/blob/7d75b6c7a0903e3f31e6ade6282f823408382f0d/packages/expo/react-native.config.js#L5-L10

The new code assumes you've done *exactly* what is listed in https://docs.expo.dev/bare/installing-expo-modules/ and that page doesn't stress how important following it *exactly* is.

In our specific case, we're in a mono repo where all relative paths are different to the RN/Expo defaults. We don't want to run node repeatedly to find the same repo root in gradle files for RN, Expo, etc so we hardcode our relative root to a constant and reference it. Here's our `settings.gradle` file:

```gradle
rootProject.name = 'AnAppName'

def relativeMonorepoRoot = '../../../..'

apply from: file("${relativeMonorepoRoot}/node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesSettingsGradle(settings)

include ':app'
includeBuild("${relativeMonorepoRoot}/node_modules/@react-native/gradle-plugin")

apply from: file("${relativeMonorepoRoot}/node_modules/expo/scripts/autolinking.gradle"); useExpoModules()
```

As a result, SDK 49 doesn't find the `apply from: new File(["node", "--print", "require.resolve('expo/package.json')"].execute(null, rootDir).text.trim(), "../scripts/autolinking.gradle")` and decides not to auto-link.

I've used `patch-package` to replace SDK 49's `react-native.config.js` with the SDK 48 version and it's all working fine now. 🥳
